lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp

```diff
@@ -390,6 +390,17 @@ void ProxyRenderDelegate::_InitRenderDelegate(MSubSceneContainer& container) {
                 break;
             }
         }
+
+        // Set the render tags to be all the tags. Vp2RenderDelegate implements
+        // render tags as a per-render item setting, so we always need to visit
+        // all the rprims to handle cases when an rprim changes from a displayed
+        // tag to a hidden tag.
```

Comment:
The performance downside of this change is for animated rprims which have a purpose which is not displayed. In that case we'll do all the updates for those rprims and then not draw the item, which is unfortunate.

We might be able to work around it by having a special case for visibility, where we update the visibility of an item first. Then, for hidden items don't bother updating anything else, just leave it all dirty.

Reply:
With this change, it sounds like purpose will only allow us to save on drawing, but not on the cost to sync changes? What's the implication for memory usage of big static scenes, time to the first frame, etc?

Reply:
I think you are right we will create all the guide objects and then not draw them. Basically we have to poll every rprim every frame to see if it goes its visibility dirtied because it's purpose changed.

I checked again how HdSt handles the render tag on an rprim changing to a hidden tag:

- HdChangeTracker::MarkRprimDirty() is called with DirtyRenderTag as one of the dirty bits, which increments rprimIndexVersion.

- In HdSt_RenderPass::_PrepareDrawItems() they detect the change in rprimIndexVersion and rebuild the list of draw items to draw (filtered by render tags).

This is what we are missing. We don't have an equivalent search over all the draw items to only update their "visibility". Items get hidden by being left out of the _drawItems list, and therefore excluded from their draw batches in _PrepareCommandBuffers.
I'm trying use our existing pass over all the rprims in execute to turn items on and off as necessary, to do that I explicitly need to consider each MRenderItem and turn on or off its visibility.

I can improve it with a combination of two changes:

- Have an early exit in Sync for rprims that are hidden by render tags, leaving the dirty bits intact so they update correctly when made visible. That prevents creating Maya stuff for guide items which are not drawn.

- Only force the pass over all render tags when that HdChangeTracker::rprimIndexVersion has changed. That prevents visiting everything and checking its visibility every refresh (I'm assuming we'd have lots of dirty + hidden stuff because we skip the update for hidden things). We'd only visit the stuff that passes the render tags test.